docs/architecture/adr-061-core-module-api.md

```diff
@@ -0,0 +1,490 @@
+# ADR 061: Core Module API
+
+## Changelog
+
+* 2022-08-18 First Draft
+
+## Status
+
+PROPOSED Not Implemented
+
+## Abstract
+
+A new core API is proposed to replace the existing `AppModule` and `sdk.Context` frameworks a set of core services
+and a new `Handler` struct. This core API aims to:
+- be simpler,
+- more extensible
+- and more stable than the current framework
+while also
+- enabling deterministic events and queries, and
+- supporting event listeners
+- and [ADR 033: Protobuf-based Inter-Module Communication](./adr-033-protobuf-inter-module-comm.md) clients.
+
+## Context
+
+Historically modules have exposed their functionality to the state machine via the `AppModule` and `AppModuleBasic`
+interfaces which have the following shortcomings:
+* both `AppModule` and `AppModuleBasic` need to be defined which is counter-intuitive
+* apps need to implement the full interfaces, even parts they don't need
+* interface methods depend heavily on unstable third party dependencies, in particular Tendermint
```

Comment:
This isn't a con IMO

Reply:
Depending on particular versions of Tendermint means you can't have a module which simultaneously supports multiple SDK versions in a hypothetical world where ABCI++ exists but maybe some users are more conservative about adopting it. More generally, it means the ecosystem needs to move in fits and starts - all modules must advance together, rather than allowing a "matrix" of compatible versions. I consider this a big con. Will try to explain this a bit more in the ADR